base_action_visibility_restriction_domain/models/ir_actions.py

```diff
@@ -0,0 +1,165 @@
+            for act in actions:
+                act_rec = self.env[act["type"]].browse(act["id"])
+                restrictions = act_rec.restriction_ids.filtered(
+                    lambda x: x.group_id.id in group_ids
+                )
+                if restrictions and any(restrictions.mapped("condition_domain")):
+                    # if at least one has domain then action is visible
+                    # and execution access will be defined in run()
+                    # otherwise action is hidden.
+                    filtered_actions.append(act)
+                elif restrictions and not any(restrictions.mapped("condition_domain")):
+                    # all restrictions are without domain. Action inaccessible.
+                    continue
+                else:
+                    # no restriction. Action accessible.
+                    filtered_actions.append(act)
+            res[binding_type] = filtered_actions
+        return res
+
+
+class IrActionsServer(models.Model):
+    _inherit = "ir.actions.server"
+
+    restriction_ids = fields.One2many("ir.actions.restriction", "server_action_id")
+
+    def run(self):
+        """
+        Prevent action running based on groups and domain
+        """
+        for action in self.sudo():
+            restrictions = action.restriction_ids.filtered(
+                lambda x: x.group_id.id in self.env.user.groups_id.ids
+            )
+            deny = False
+            rec_ids = self.env.context.get("active_id") or self.env.context.get(
+                "active_ids"
+            )
+            records = self.env[self.model_name].browse(rec_ids)
+            blocked_records = False
+            for restr in restrictions:
+                if restr.condition_domain not in ["[]", False, ""]:
+                    # domain allows to run action
+                    filtered_records = records.filtered_domain(
+                        safe_eval(restr.condition_domain)
+                    )
+                    blocked_records = records - filtered_records
+                    if not blocked_records:
+                        # all records available for action
+                        continue
+                    else:
+                        # some records are blocked - raise error
+                        deny = True
+                        break
+                else:
+                    # no domain - block
+                    deny = True
+                    break
+            if deny:
+                if blocked_records:
+                    raise AccessError(
+                        _(
+                            "You don't have enough access rights to run "
+                            "this action on records: %s." % blocked_records
+                        )
+                    )
+                else:
+                    raise AccessError(
+                        _("You don't have enough access rights to run this action.")
+                    )
```

Comment:
**suggestion (non-blocking)**: this could be slightly more readable. Since you set `deny` and immediately `break` to check `deny`, we could do without `deny` entirely IMO

```suggestion
            rec_ids = self.env.context.get("active_id") or self.env.context.get(
                "active_ids"
            )
            records = self.env[self.model_name].browse(rec_ids)
            for restr in restrictions:
                if restr.condition_domain in ["[]", False, ""]:
                    # no domain - block
                    raise AccessError(
                        _("You don't have enough access rights to run this action.")
                    )
                else:
                    # domain allows to run action
                    filtered_records = records.filtered_domain(
                        safe_eval(restr.condition_domain)
                    )
                    blocked_records = records - filtered_records
                    if blocked_records:
                        raise AccessError(
                            _(
                                "You don't have enough access rights to run "
                                "this action on records: %s." % blocked_records
                            )
                        )
```
At that point you could even drop the `else` clause entirely.